File: Tests/Scraping/FormScrapingTests.m
//  FormScrapingTests.m
//
//  Copyright 2013 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

#import "AwfulScrapingTestCase.h"
#import "AwfulForm.h"

@interface FormScrapingTests : AwfulScrapingTestCase

@end

@implementation FormScrapingTests

- (AwfulForm *)scrapeFormFromFixtureNamed:(NSString *)fixtureName
{
    HTMLDocument *document = LoadFixtureNamed(fixtureName);
    HTMLElement *formElement = [document firstNodeMatchingSelector:@"form[name='vbform']"];
    AwfulForm *form = [[AwfulForm alloc] initWithElement:formElement];
    [form scrapeThreadTagsIntoManagedObjectContext:self.managedObjectContext];
    NSError *error;
    BOOL ok = [self.managedObjectContext save:&error];
    NSAssert(ok, @"error saving context after scraping thread tags: %@", error);
    return form;
}

- (void)testReply
{
    AwfulForm *form = [self scrapeFormFromFixtureNamed:@"newreply"];
    XCTAssertEqual(form.threadTags.count, (NSUInteger)0);
    NSMutableDictionary *parameters = [form recommendedParameters];
    XCTAssertEqualObjects(parameters[@"action"], @"postreply");
    XCTAssertEqualObjects(parameters[@"threadid"], @"3507451");
    XCTAssertEqualObjects(parameters[@"formkey"], @"0253d85a945b60daa0165f718df82b8a");
    XCTAssertEqualObjects(parameters[@"form_cookie"], @"80c74b48f557");
    XCTAssertNotNil(parameters[@"message"]);
    XCTAssertNotEqual([parameters[@"message"] rangeOfString:@"terrible"].location, (NSUInteger)NSNotFound);
    XCTAssertNotNil(parameters[@"parseurl"]);
    XCTAssertNotNil(parameters[@"bookmark"]);
    XCTAssertNil(parameters[@"disablesmilies"]);
    XCTAssertNil(parameters[@"signature"]);
}

- (void)testReplyWithAmazonSearch
{
    AwfulForm *form = [self scrapeFormFromFixtureNamed:@"newreply-amazon-form"];
    XCTAssertNotNil([form recommendedParameters][@"threadid"]);
}

- (void)testThread
{
    AwfulForm *form = [self scrapeFormFromFixtureNamed:@"newthread"];
    XCTAssertEqual(form.threadTags.count, (NSUInteger)51);
    NSArray *allThreadTags = [AwfulThreadTag fetchAllInManagedObjectContext:self.managedObjectContext];
    XCTAssertEqual(allThreadTags.count, form.threadTags.count);
    XCTAssertTrue(form.secondaryThreadTags.count == 0);
    NSDictionary *parameters = form.allParameters;
    XCTAssertNotNil(parameters[@"subject"]);
    XCTAssertNotNil(parameters[@"message"]);
    XCTAssertEqualObjects(parameters[@"forumid"], @"1");
    XCTAssertEqualObjects(parameters[@"action"], @"postthread");
    XCTAssertEqualObjects(parameters[@"formkey"], @"0253d85a945b60daa0165f718df82b8a");
    XCTAssertEqualObjects(parameters[@"form_cookie"], @"e29a15add831");
    XCTAssertEqualObjects(parameters[@"parseurl"], @"yes");
    XCTAssertEqualObjects(parameters[@"bookmark"], @"yes");
}

- (void)testAskTellThread
{
    AwfulForm *form = [self scrapeFormFromFixtureNamed:@"newthread-at"];
    XCTAssertEqual(form.threadTags.count, (NSUInteger)55);
    XCTAssertEqual(form.secondaryThreadTags.count, (NSUInteger)2);
    NSArray *secondaryTags = [AwfulThreadTag fetchAllInManagedObjectContext:self.managedObjectContext
                                                    matchingPredicateFormat:@"imageName IN { 'ama', 'tma' }"];
    XCTAssertEqual(secondaryTags.count, form.secondaryThreadTags.count);
}

- (void)testSAMartThread
{
    AwfulForm *form = [self scrapeFormFromFixtureNamed:@"newthread-samart"];
    XCTAssertEqual(form.threadTags.count, (NSUInteger)69);
    XCTAssertEqual(form.secondaryThreadTags.count, (NSUInteger)4);
    NSArray *secondaryTags = [AwfulThreadTag fetchAllInManagedObjectContext:self.managedObjectContext
                                                    matchingPredicateFormat:@"imageName LIKE 'icon*ing'"];
    NSIndexSet *secondaryIndexes = [secondaryTags indexesOfObjectsPassingTest:^BOOL(AwfulThreadTag *threadTag, NSUInteger i, BOOL *stop) {
        return threadTag.threadTagID.integerValue < 5;
    }];
    XCTAssertEqual(secondaryIndexes.count, form.secondaryThreadTags.count);
}

- (void)testMessage
{
    AwfulForm *form = [self scrapeFormFromFixtureNamed:@"private-reply"];
    NSDictionary *parameters = form.allParameters;
    XCTAssertNotNil(parameters[@"message"]);
    XCTAssertNotEqual([parameters[@"message"] rangeOfString:@"InFlames235 wrote"].location, (NSUInteger)NSNotFound);
}

@end

File: Tests/Scraping/FormScrapingTests.swift
//  FormScrapingTests.swift
//
//  Copyright 2013 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

import XCTest
import Awful

class FormScrapingTests: ScrapingTestCase {

    private func scrapeFormFixtureNamed(fixtureName: String) -> AwfulForm {
        let document = loadFixtureNamed(fixtureName)
        let formElement = document.firstNodeMatchingSelector("form[name='vbform']")
        let form = AwfulForm(element: formElement)
        form.scrapeThreadTagsIntoManagedObjectContext(managedObjectContext)
        var error: NSError?
        let ok = managedObjectContext.save(&error)
        assert(ok, "error saving context after scraping thread tags: \(error!)")
        return form
    }

    func testReply() {
        let form = scrapeFormFixtureNamed("newreply")
        XCTAssertTrue(form.threadTags == nil)
        let parameters = form.recommendedParameters()
        XCTAssertEqual(parameters["action"] as String, "postreply")
        XCTAssertEqual(parameters["threadid"] as String, "3507451")
        XCTAssertEqual(parameters["formkey"] as String, "0253d85a945b60daa0165f718df82b8a")
        XCTAssertEqual(parameters["form_cookie"] as String, "80c74b48f557")
        XCTAssertTrue((parameters["message"] as String).rangeOfString("terrible") != nil)
        XCTAssertNotNil(parameters["parseurl"])
        XCTAssertNotNil(parameters["bookmark"])
        XCTAssertNil(parameters["disablesmilies"])
        XCTAssertNil(parameters["signature"])
    }

    func testReplyWithAmazonSearch() {
        let form = scrapeFormFixtureNamed("newreply-amazon-form")
        XCTAssertNotNil(form.recommendedParameters()["threadid"])
    }

    func testThread() {
        let form = scrapeFormFixtureNamed("newthread")
        XCTAssertTrue(form.threadTags.count == 51)
        XCTAssertEqual(ThreadTag.numberOfObjectsInManagedObjectContext(managedObjectContext), form.threadTags.count)
        XCTAssertTrue(form.secondaryThreadTags == nil)
        let parameters = form.allParameters
        XCTAssertNotNil(parameters["subject"])
        XCTAssertNotNil(parameters["message"])
        XCTAssertEqual(parameters["forumid"] as String, "1")
        XCTAssertEqual(parameters["action"] as String, "postthread")
        XCTAssertEqual(parameters["formkey"] as String, "0253d85a945b60daa0165f718df82b8a")
        XCTAssertEqual(parameters["form_cookie"] as String, "e29a15add831")
        XCTAssertEqual(parameters["parseurl"] as String, "yes")
        XCTAssertEqual(parameters["bookmark"] as String, "yes")
    }
    
    func testAskTellThread() {
        let form = scrapeFormFixtureNamed("newthread-at")
        XCTAssertTrue(form.threadTags.count == 55)
        XCTAssertTrue(form.secondaryThreadTags.count == 2)
        let secondaryTags = ThreadTag.fetchAllInManagedObjectContext(managedObjectContext, matchingPredicate: NSPredicate(format: "imageName IN { 'ama', 'tma' }"))
        XCTAssertEqual(secondaryTags.count, form.secondaryThreadTags.count);
    }
    
    func testSAMartThread() {
        let form = scrapeFormFixtureNamed("newthread-samart")
        XCTAssertTrue(form.threadTags.count == 69)
        XCTAssertTrue(form.secondaryThreadTags.count == 4)
        let possibleSecondaryTags = ThreadTag.fetchAllInManagedObjectContext(managedObjectContext, matchingPredicate: NSPredicate(format: "imageName LIKE 'icon*ing'")) as [ThreadTag]
        let secondaryTags = possibleSecondaryTags.filter { $0.threadTagID!.toInt()! < 5 }
        XCTAssertEqual(secondaryTags.count, form.secondaryThreadTags.count)
    }
    
    func testMessage() {
        let form = scrapeFormFixtureNamed("private-reply")
        let parameters = form.allParameters
        XCTAssertNotNil(parameters["message"])
        XCTAssertTrue((parameters["message"] as String).rangeOfString("InFlames235 wrote") != nil)
    }
}
